orttraining/orttraining/test/optimizer/graph_transformer_utils_test.cc

```diff
@@ -20,25 +20,36 @@ TEST(GraphTransformerUtilsTestsForTraining, TestGenerateGraphTransformers) {
   std::string l1_rule1 = "EliminateIdentity";
   std::string l1_transformer = "ConstantFolding";
   std::string l2_transformer = "ConvActivationFusion";
-  std::vector<std::string> custom_list = {l1_rule1, l1_transformer, l2_transformer};
+  std::unordered_set<std::string> disabled = {l1_rule1, l1_transformer, l2_transformer};
+  CPUExecutionProvider cpu_ep(CPUExecutionProviderInfo{});
 
-  auto transformers = training::transformer_utils::GenerateTransformers(TransformerLevel::Level1, {}, {}, custom_list);
-  ASSERT_TRUE(transformers.size() == 1);
+  auto all_transformers = optimizer_utils::GenerateTransformers(TransformerLevel::Level1, {}, cpu_ep);
+  auto filtered_transformers = optimizer_utils::GenerateTransformers(TransformerLevel::Level1, {}, cpu_ep, disabled);
 
+  // check ConstantFolding transformer was removed
+  ASSERT_TRUE(filtered_transformers.size() == all_transformers.size() - 1);
+
+  // check EliminateIdentity rule was removed from inside the rule based transformer
   auto l1_rule_transformer_name = optimizer_utils::GenerateRuleBasedTransformerName(TransformerLevel::Level1);
   RuleBasedGraphTransformer* rule_transformer = nullptr;
-  for (const auto& transformer : transformers) {
+  for (const auto& transformer : filtered_transformers) {
     if (transformer->Name() == l1_rule_transformer_name) {
-      rule_transformer = dynamic_cast<RuleBasedGraphTransformer*>(transformers[0].get());
+      rule_transformer = static_cast<RuleBasedGraphTransformer*>(transformer.get());
+
+      // get the full set of rules and check EliminateIdentity was correctly removed
+      auto l1_rewrite_rules = optimizer_utils::GenerateRewriteRules(TransformerLevel::Level1);
+      ASSERT_TRUE(rule_transformer->RulesCount() == l1_rewrite_rules.size() - 1);
+      break;
     }
   }
-  ASSERT_TRUE(rule_transformer && rule_transformer->RulesCount() == 1);
 
-  transformers = training::transformer_utils::GenerateTransformers(TransformerLevel::Level2, {}, {}, custom_list);
+  ASSERT_TRUE(rule_transformer) << "RuleBased transformer should have been added by GenerateTransformers";
+
```

Comment:
Move this after line 37? Actualy can this ever be null?

Reply:
It's null if we didn't add a rule based transformer to the list in GenerateTransformers. The 'rule' we're matching (EliminateIdentity) is inside of a transformer called L1_RuleBasedTransformer.